### PR TITLE
use poetry insted of Pipfile

### PR DESCRIPTION
**Pull request by RCheese**
#123

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -95,6 +95,7 @@ The generator (cookiecutter) will ask you for some data, you might want to have
 The input variables, with their default values (some auto generated) are:
 
 * `project_name`: The name of the project
+* `project_version`: The version of the project
```
Comment (tiangolo) [inline README.md:98]:
```suggestion
```

I would prefer to have a default fixed version `0.1.0` for all new generated projects and let everyone customize ii afterwards if needed.

File: cookiecutter.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
     "project_name": "Base Project",
     "project_slug": "{{ cookiecutter.project_name|lower|replace(' ', '-') }}",
+    "project_version": "0.1.0",
```
Comment (tiangolo) [inline cookiecutter.json:4]:
```suggestion
```

I would prefer to leave a default `0.1.0` and let users customize afterwards if needed.

File: {{cookiecutter.project_slug}}/backend/backend.dockerfile [hunk 3]
```diff
@@ -1,6 +1,7 @@
 FROM tiangolo/uvicorn-gunicorn-fastapi:python3.7
 
-RUN pip install celery~=4.3 passlib[bcrypt] tenacity requests emails "fastapi>=0.47.0" "uvicorn>=0.11.1" gunicorn pyjwt python-multipart email_validator jinja2 psycopg2-binary alembic SQLAlchemy email_validator
+ADD requirements.txt .
+RUN pip install -r requirements.txt
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/backend.dockerfile:4]: If doing this, I would prefer to add Poetry to the Docker images and install with Poetry without creating an env inside the Dockerfiles instead of having a pre-build step outside of Docker that depends on Poetry installed locally.

File: {{cookiecutter.project_slug}}/backend/celeryworker.dockerfile [hunk 4]
```diff
@@ -1,6 +1,7 @@
 FROM python:3.7
 
-RUN pip install raven celery~=4.3 passlib[bcrypt] tenacity requests "fastapi>=0.16.0" emails pyjwt email_validator jinja2 psycopg2-binary alembic SQLAlchemy email_validator
+ADD requirements.txt .
+RUN pip install -r requirements.txt
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/celeryworker.dockerfile:4]: The same here, I would prefer to add Poetry to the Docker images and install with Poetry without creating an env inside the Dockerfiles instead of having a pre-build step outside of Docker that depends on Poetry installed locally.

File: {{cookiecutter.project_slug}}/backend/make-requirements.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+poetry export -f requirements.txt -o requirements.txt $(if [[ "$1" == "--dev" ]]; then echo "--dev"; fi)
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/make-requirements.sh:3]: I would prefer not to have this file that requires Poetry installed locally to work and instead add poetry to the Dockerfiles.

File: {{cookiecutter.project_slug}}/backend/poetry.lock [hunk 6]
```diff
@@ -0,0 +1,2204 @@
+[[package]]
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/poetry.lock:1]: Please remove the lock, as it could end up varying a lot depending on the system (Windows, Linux, etc). I would prefer to let users commit it if they want to, but not to have a default one there already.

File: {{cookiecutter.project_slug}}/backend/pyproject.toml [hunk 7]
```diff
@@ -1,19 +1,11 @@
-[[source]]
-name = "pypi"
-url = "https://pypi.org/simple"
-verify_ssl = true
+[tool.poetry]
+name = "{{cookiecutter.project_slug}}"
+version = "{{cookiecutter.project_version}}"
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/pyproject.toml:3]:
```suggestion
name = "backend"
version = "0.1.0"
```

Let's have this file without template variables so that it can work as is. This might also help bots reading it, e.g. Dependabot.

File: {{cookiecutter.project_slug}}/backend/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,629 @@
+alembic==1.4.2 \
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/requirements.txt:1]: Please remove this file from git.

File: {{cookiecutter.project_slug}}/backend/tests.dockerfile [hunk 9]
```diff
@@ -1,6 +1,7 @@
 FROM python:3.7
 
-RUN pip install requests pytest tenacity passlib[bcrypt] "fastapi>=0.16.0" psycopg2-binary SQLAlchemy email_validator
+ADD requirements.txt .
+RUN pip install -r requirements.txt
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/backend/tests.dockerfile:4]: Here we would also have Poetry installed and using the `pyproject.toml` file.

File: {{cookiecutter.project_slug}}/cookiecutter-config-file.yml [hunk 10]
```diff
@@ -1,5 +1,6 @@
 default_context:
   project_name: '{{ cookiecutter.project_name }}'
+  project_version: '{{ cookiecutter.project_version }}'
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/cookiecutter-config-file.yml:3]:
```suggestion
```

Please remove the project version, that would be a default of `0.1.0`.

File: {{cookiecutter.project_slug}}/docker-compose.shared.base-images.yml [hunk 11]
```diff
@@ -1,7 +1,7 @@
 version: '3.3'
 services:
   db:
-    image: postgres:11
+    image: postgres:12
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/docker-compose.shared.base-images.yml:4]:
```suggestion
    image: postgres:11
```

Let's leave the upgrade to a separate PR to keep the scope clear and narrow.

File: {{cookiecutter.project_slug}}/scripts/prepare.sh [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#! /usr/bin/env sh
```
Comment (tiangolo) [inline {{cookiecutter.project_slug}}/scripts/prepare.sh:1]: Please remove this script, as the installation would be part of the Dockerfiles.

File: {{cookiecutter.project_slug}}/cookiecutter-config-file.yml (hunk 10 above)
Comment (RCheese) [reply {{cookiecutter.project_slug}}/cookiecutter-config-file.yml:3]: Project version need to poetry pyproject.toml. I can fix it in this file. It's ok?

File: {{cookiecutter.project_slug}}/backend/backend.dockerfile (hunk 3 above)
Comment (RCheese) [reply {{cookiecutter.project_slug}}/backend/backend.dockerfile:4]: Then I'll use multi-build stage in Dockerfile, alright?

**Comment by RCheese** [comment]
@tiangolo 
1. https://python-poetry.org/docs/#installing-with-pip
2. What a reason for use poetry inside deployed container?

**Comment by tiangolo** [comment]
Thanks for your contribution! :cake: I updated it a bit and it's now merged. :rocket: :tada:
